pymonetdb/mapi.py

```diff
@@ -295,8 +304,24 @@ def _challenge_response(self, challenge):
             raise NotSupportedError("Unsupported hash algorithms required"
                                     " for login: %s" % hashes)
 
-        return ":".join(["BIG", self.username, pwhash, self.language,
-                         self.database]) + ":"
+        response = ":".join(["BIG", self.username, pwhash, self.language, self.database]) + ":"
+
+        if len(challenges) >= 7:
+            options_level = 0
+            for part in challenges[6].split(","):
+                if part.startswith("sql="):
+                    try:
+                        options_level = int(part[4:])
+                    except ValueError:
+                        raise OperationalError("invalid sql options level in server challenge: " + part)
+            options = []
+            for opt in self.handshake_options:
+                if opt.level < options_level:
+                    options.append(opt.name + "=" + str(int(opt.value)))
+                    opt.sent = True
```

Comment:
ah ok, you modify the object here so you don't want to modify the original.

Reply:
but then the iteration of the list doesn't make a copy, it just makes a new list of the same items. you need to use deepcopy for that. But the list is initiated during connection init, so there is no need to copy anyway since it is unique per connection?